docs/concepts.rst

```diff
@@ -116,6 +116,46 @@ DAGs can be used as context managers to automatically assign new operators to th
 
     op.dag is dag # True
 
+.. _concepts:functional_dags:
+
+Functional DAGs
```

Comment:
Random thought. Functional dags is perhaps not the right name for this whole feature. It's not like functional programming where a something is run without side-effects.

What would people think about renaming this to "Literate DAGs"? I'll take to the mailing list if people agree, just something doesn't quite sit right with me about "functional" today.

If no one else likes this/thinks functional is great then I'll not worry.

Reply:
And we can discuss this _after_ this PR is merged if we want. It's not a blocker on merging the PR.

Reply:
That's fair. The Functional DAGs comes from the AIP title itself. I do understand the confusion, but not sure if literate is better. 

Ideally though, having XComArg and all this, it should allow DAGs to be written without as much side effects. You will be able to paramatrize an operator such that it is indeed more functional (takes A and B, outputs C) instead of current approach where you mostly generate artifacts that depend on Airflow context and such they are side effects.

Not sure if the above makes sense. In my mind, it makes it more functional (less prone to side effects) as you make operators resemble more functions, and XComArgs map to function parameters and function outputs.

Reply:
Does this need any name? Can't we just say that this is a way how users can define DAGs? Personally I would encourage users to use this way instead of using XComs. WDYT?

Reply:
Fair. Main issue is that it does introduce a new syntax so for backwards compatible issues I would add a name to this.

We can switch all documentation to use this style on 2.0 and therefore remove this specific section (or collapse it)

Reply:
@casassg do we have any backwards issues other than metaclass?

Reply:
That's true. Mostly saying as a full refactor of user docs on how to write DAGs may be more adequate for Airflow 2.0 than 1.10. But happy to reconsider this. If we go down that path, we should do it in a separate PR though. This PR is already become large enough

Reply:
Should we make an issue for that, discuss and then adjust the docs in a follow-up PR?

Reply:
Created #9415